// @tarojs/mini-runner/dist/webpack/base.conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Chain = require("webpack-chain");
const runner_utils_1 = require("@tarojs/runner-utils");
exports.default = (appPath) => {
    const chain = new Chain();
    chain.merge({
        resolve: {
            extensions: ['.js', '.jsx', '.ts', '.tsx', '.mjs', '.vue'],
            mainFields: ['browser', 'module', 'main'],
            symlinks: true,
            modules: [
                'node_modules',
                path.join(appPath, 'node_modules')
            ],
            alias: {
                // 小程序使用 regenerator-runtime@0.11
                'regenerator-runtime': require.resolve('regenerator-runtime')
            }
        },
        resolveLoader: {
            modules: [
                'node_modules'
            ]
        },
        optimization: {
            sideEffects: true
        }
    });
    chain.resolve
        .plugin('MultiPlatformPlugin')
        .use(runner_utils_1.MultiPlatformPlugin, ['described-resolve', 'resolve']);
    return chain;
};
